### PR TITLE
[autoparallel] add data_ptr

### PR DESCRIPTION
**Pull request by super-dainiu**
### What's new?
Since all data in `OperationData` are meta tensors, their `tensor.data_ptr()=0`. I replace the original `tensor.data_ptr()` with a customized function generated by `uuid4()`.

### Why data_ptr?
Generally speaking, for fine-grained memory calculations, I need to specify which tensors are the same tensors which occupy the same piece of memory on the devices.

### Test
image

### CODE REVIEW

File: colossalai/auto_parallel/meta_profiler/metainfo.py [hunk 1]
```diff
@@ -80,9 +81,10 @@ def compute_sharded_tensor(self, operation_data: OperationData, sharding_spec: S
             else:
                 # sharded according to device_mesh shape
                 new_shape.append(dim // np.prod(np.array([device_mesh.mesh_shape[i] for i in shard.shard_list])))
-
+        new_data = torch.zeros(new_shape, device="meta")
+        new_data.data_ptr = operation_data.data.data_ptr
         return OperationData(name=operation_data.name,
-                             data=torch.zeros(new_shape, device="meta"),
+                             data=new_data,
```
Comment (super-dainiu) [inline colossalai/auto_parallel/meta_profiler/metainfo.py:87]: Each new data fed into `OperationData` should be marked with `data_ptr`

File: colossalai/auto_parallel/tensor_shard/node_handler/__init__.py [hunk 2]
```diff
@@ -2,13 +2,13 @@
...
 from .conv_handler import ConvFunctionHandler, ConvModuleHandler
-from .getatrr_handler import GetattrHandler
+from .getattr_handler import GetattrHandler
 from .layer_norm_handler import LayerNormModuleHandler
 from .linear_handler import LinearFunctionHandler, LinearModuleHandler
 from .matmul_handler import MatMulHandler
 from .normal_pooling_handler import NormPoolingHandler
 from .output_handler import OuputHandler
-from .placeholder_handler import PlacehodlerHandler
+from .placeholder_handler import PlaceholderHandler
```
Comment (super-dainiu) [inline colossalai/auto_parallel/tensor_shard/node_handler/__init__.py:11]: typo

File: colossalai/auto_parallel/tensor_shard/sharding_strategy.py [hunk 3]
```diff
@@ -14,6 +15,17 @@
 __all__ = ['OperationDataType', 'OperationData', 'TrainCycleItem', 'MemoryCost', 'ShardingStrategy', 'StrategiesVector']
 
 
+def _assert_non_trivial_data_ptr(data: Union[torch.Tensor, List, Tuple]) -> None:
```
Comment (super-dainiu) [inline colossalai/auto_parallel/tensor_shard/sharding_strategy.py:18]: This assert every tensor has `data_ptr()` that returns other than `0`

File: colossalai/auto_parallel/tensor_shard/sharding_strategy.py [hunk 4]
```diff
@@ -43,8 +55,10 @@ class OperationData:
 
     def __post_init__(self):
         # if no logical shape is specified, use the data shape as the logical shape
-        if self.logical_shape is None and isinstance(self.data, torch.Tensor):
-            self.logical_shape = self.data.shape
+        if self.logical_shape is None:
```
Comment (super-dainiu) [inline colossalai/auto_parallel/tensor_shard/sharding_strategy.py:58]: Logic shape should have tree structure if data also has tree structure

**Comment by FrankLeeeee** [comment]
This is still a draft?

File: colossalai/auto_parallel/meta_profiler/metainfo.py (hunk 1 above)
Comment (FrankLeeeee) [reply colossalai/auto_parallel/meta_profiler/metainfo.py:87]: What if there are two tensors which share the same underlying data?

File: colossalai/auto_parallel/tensor_shard/node_handler/binary_elementwise_handler.py [hunk 5]
```diff
@@ -46,6 +47,9 @@ def _get_arg_value(idx):
                 # as it won't affect the strategy generation
                 assert isinstance(self.node.args[idx], (int, float))
                 meta_data = torch.Tensor([self.node.args[idx]]).to('meta')
+                # (super-dainiu): each newly created tensor on meta device will have a unique ``data_ptr()``
+                data_ptr = uuid.uuid4()
+                meta_data.data_ptr = lambda: data_ptr
```
Comment (FrankLeeeee) [inline colossalai/auto_parallel/tensor_shard/node_handler/binary_elementwise_handler.py:52]: why assign a lambda function to data_ptr?

File: colossalai/fx/proxy.py [hunk 6]
```diff
@@ -1,12 +1,39 @@
...
+    """
+    Set a non-trivial data_ptr for the data if it is a ``torch.Tensor`` and is on ``meta`` device
+    """
+    if isinstance(data, list) or isinstance(data, tuple) and not isinstance(data, torch.Size):
+        for item in data:
+            _set_data_ptr(item)
+    elif isinstance(data, torch.Tensor):
+        if not data.data_ptr():
+            data_ptr = uuid.uuid4()
+            data.data_ptr = lambda: data_ptr
```
Comment (FrankLeeeee) [inline colossalai/fx/proxy.py:23]: Use PyTree instead to set data pointer instead as different data structures can be expected.

**Pull request closed without merging**